components/brave_wallet/browser/zcash/zcash_orchard_storage.cc

```diff
@@ -0,0 +1,438 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_wallet/browser/zcash/zcash_orchard_storage.h"
+
+#include "base/files/file_util.h"
+#include "base/strings/stringprintf.h"
+#include "brave/components/brave_wallet/common/hex_utils.h"
+#include "sql/statement.h"
+#include "sql/transaction.h"
+
+namespace brave_wallet {
+
+namespace {
+#define kNotesTable "notes"
+#define kSpentNotesTable "spent_notes"
+#define kAccountMeta "account_meta"
+
+const int kEmptyDbVersionNumber = 1;
+const int kCurrentVersionNumber = 2;
+
+std::optional<uint32_t> ReadUint32(sql::Statement& statement, size_t index) {
+  auto v = statement.ColumnInt64(index);
+  if (v < 0 || v > std::numeric_limits<uint32_t>::max()) {
+    return std::nullopt;
+  }
+  return static_cast<uint32_t>(v);
+}
+
+}  // namespace
+
+ZCashOrchardStorage::ZCashOrchardStorage(base::FilePath path_to_database)
+    : db_file_path_(path_to_database), database_() {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+}
+
+ZCashOrchardStorage::~ZCashOrchardStorage() {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+}
+
+bool ZCashOrchardStorage::EnsureDbInit() {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+  if (database_.is_open()) {
+    return true;
+  }
+  return CreateOrUpdateDatabase();
+}
+
+void ZCashOrchardStorage::ResetDatabase() {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+
+  database_.Close();
+  sql::Database::Delete(db_file_path_);
+}
+
+bool ZCashOrchardStorage::CreateOrUpdateDatabase() {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+
+  const base::FilePath dir = db_file_path_.DirName();
+  if (!base::DirectoryExists(dir) && !base::CreateDirectory(dir)) {
+    NOTREACHED_IN_MIGRATION();
+    return false;
+  }
+
+  if (!database_.Open(db_file_path_)) {
+    NOTREACHED_IN_MIGRATION();
+    return false;
+  }
+
+  sql::MetaTable meta_table;
+  if (!meta_table.Init(&database_, kEmptyDbVersionNumber,
+                       kEmptyDbVersionNumber)) {
+    NOTREACHED_IN_MIGRATION();
+    database_.Close();
+    return false;
+  }
+
+  if (meta_table.GetVersionNumber() == kEmptyDbVersionNumber) {
+    if (!CreateSchema() ||
+        !meta_table.SetVersionNumber(kCurrentVersionNumber)) {
+      NOTREACHED_IN_MIGRATION();
+      database_.Close();
+      return false;
+    }
+  } else if (meta_table.GetVersionNumber() < kCurrentVersionNumber) {
+    if (!UpdateSchema() ||
+        !meta_table.SetVersionNumber(kCurrentVersionNumber)) {
+      NOTREACHED_IN_MIGRATION();
+      database_.Close();
+      return false;
+    }
+  }
+
+  return true;
+}
+
+bool ZCashOrchardStorage::CreateSchema() {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+
+  sql::Transaction transaction(&database_);
+  return transaction.Begin() &&
+         database_.Execute("CREATE TABLE " kNotesTable
```

Comment:
nit: it seems like it might be useful to parameterize these table names rather than just concatenating them as an extra layer of safety when building these SQL statements.

Reply:
https://github.com/brave/brave-core/pull/24547#discussion_r1669849182
Looks like approach which uses "define" is more commonly used in chromium.

Reply:
SGTM, consistency is better than doing it different ways IMO.